code.js

```diff
@@ -1,3 +1,49 @@
-function tsp_hk(distance_matrix) {
-    return -1;
-}
+function tsp_hk(dist) {
+    // cities is the set of cities not visited so far, including start
+    let cities = [];
+    for (let i = 0; i < dist.length; i++) {
+        cities.push(i);
+    }
+    let best = Infinity;
+
+    //memoization
+    let cache = [];
+
+    for (let i = 0; i < cities.length; i++) {
+        let temp = heldKarp(cities, i);
+        if (temp < best) { best = temp; }
+    }
+    return best;    
+
+
+    function heldKarp(cities, start) {
+        let key = JSON.stringify([cities, start]);
```

Comment:
What if `cities` is in a different order?

Reply:
I could be wrong, but I can't think of any case where cities would not be already sorted.

Reply:
Ok, but you'll have to convince me that this is the case.